### ISSUE TITLE
Carbon component to use <time> attribute

### ISSUE THREAD

**Issue by imliam** [blade-ui-kit/blade-ui-kit#26]
Is there a reason why the `<x-carbon>` component doesn't generate a `<time>` HTML element under the hood? It seems like the appropriate semantic element to use, rather than a plain `<span>` - figure there must be some reason behind the choice?

**Comment by bjuppa** [comment]
Are you referring to this line @imliam? https://github.com/blade-ui-kit/blade-ui-kit/blob/1baaa2d04555e7060c70dc22c29d441f37dff689/resources/views/components/date-time/carbon.blade.php#L27

I think I agree with you, that `<span>` could possibly be a `<time>`, like [line 3 in the same view](https://github.com/blade-ui-kit/blade-ui-kit/blob/1baaa2d04555e7060c70dc22c29d441f37dff689/resources/views/components/date-time/carbon.blade.php#L3)?

**Comment by imliam** [comment]
Didn't even realise it uses `<time>` there for the human-readable format, the docs also show `<span>` for its example currently.

But yes, that's exactly what I mean in that they could both be `<time>` elements.

**Comment by bjuppa** [comment]
I'm glad the `datetime` attribute may not only contain atom-format strings as I initially thought!  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/time
says the current default format in the component `Y-m-d H:i:s` 
is ok, along with many others, but not any format...

I think this component may benefit from splitting the `$format` prop into one specifically for the `datetime` attribute of `<time>`, and another format for what we want the user to read. What do you think?

**Comment by bjuppa** [comment]
I was involved in creating a similiar carbon "component" (although plain Blade `@include` or old-style `@component` for compatibility with older Laravel versions).
- https://github.com/kontenta/kontour/blob/master/docs/time-templates.md
- https://github.com/kontenta/kontour/blob/master/resources/views/elements/time.blade.php

That example also handles instances of `Carbon\CarbonInterval`, but maybe that would be better suited for a separate `carbon-interval` component?

Also, the example handles `$slot` to allow passing any html contents into the readable part. Could be nice here too?

**Comment by bjuppa** [comment]
And... just one more today I promise 😄 

Do we need to require `DateTimeInterface` for the `$date` prop? It would be nice to pass strings too for parsing? Or is it better to require the caller to do the parsing?
But if that is the case, maybe we should require the caller to do the formatting for `datetime` too? 😬 

https://github.com/blade-ui-kit/blade-ui-kit/blob/1baaa2d04555e7060c70dc22c29d441f37dff689/src/Components/DateTime/Carbon.php#L30

**Comment by abenettt** [comment]
> I think this component may benefit from splitting the `$format` prop into one specifically for the `datetime` attribute of `<time>`, and another format for what we want the user to read. What do you think?

I don't think this would be necessary actually as the `datetime` attribute is not userfacing and "only" there for the semantic correctness.


> Do we need to require `DateTimeInterface` for the `$date` prop? It would be nice to pass strings too for parsing? Or is it better to require the caller to do the parsing?

I think we should keep the requiring the `DateTimeInterface` as the user presumably is already working with an instance of `DateTimeInterface` themselves. There is no need for the user to create a string out of that only for us to parse it back into a `DateTimeInterface` instance.

**Comment by bjuppa** [comment]
> I don't think this would be necessary actually as the `datetime` attribute is not userfacing and "only" there for the semantic correctness.

I guess it depends on what resolution you're looking for, and that could be on a case by case basis. If you're displaying a date with a time you'd want the `datetime` to include both (and that could be in atom-format). But if displaying just a date *without* time, like "September 3" to the user, you want the `datetime` at lower resolution, for example `Y-m-d` so that any programatic access knows it references a full day.

I think `datetime` is for more than just semantic correctness, it can be used for programatic stuff by others parsing your html, like say some calendar plugin (or a scraper 😬). But it can also be used to enhance your own interface! For example in this component we could let the javascript read from the actual `datetime` attribute to convert it into the user's local timezone and update the `innerHTML` of the very same `<time>` element!

Hm... I haven't played enough with Alpine to know if one can easily observe a specific attribute of an element in the DOM for changes. Is that doable?

> I think we should keep the requiring the `DateTimeInterface` as the user presumably is already working with an instance of `DateTimeInterface` themselves. There is no need for the user to create a string out of that only for us to parse it back into a `DateTimeInterface` instance.

We could have it accept a string *or* a `DateTimeInterface` for flexibility, by removing the typehint. The downside is that the API becomes less clear... unfortunately we won't have union types until PHP8!

**Comment by bjuppa** [comment]
I can see in the file history this was originally called `date-time` but changed to `carbon`... presumably for clarifying that it operates on `Carbon`(-ish) objects only. But as it handles any `DateTimeInterface` maybe the name could be changed back? 😄 On the other hand, if we want it to work for dates *without* time-part, `date-time` is still not a good component-name.

`CarbonInterval` extends PHPs native `DateInterval` so even if we add that too as an option we could keep the non-carbon naming.

What about four components? Just thinking out loud here...
- `date-time`
- `date`
- `time`
- `time-interval`

...or as our component essentially adds functionality to HTML's `<time>` in all the above use-cases, why not use `<x-time>`?

**Comment by abenettt** [comment]
I like the idea of the four components, it makes what is supposed to happen more explicit.
Also, as we're a laravel package it can be presumed that the user is using `Carbon` or `CarbonImmutable` so we could require those (maybe the `CarbonInterface`) instead of the more generic `DateTimeInterface`. This would allow use easier access to the `Carbon` functions.

**Comment by abenettt** [comment]
Using separate components also lets us implicitly know the format of the `datetime` property. We might want to allow the user to override the format but a default one can be set.

**Comment by bjuppa** [comment]
The longer I think about this, the more I lean towards wanting a single component `<x-time>` to handle all cases of date-time, date, time as well as time-intervals 😄

**Comment by driesvints** [comment]
Sorry for the lack of response here btw. I'm hoping to get more thoroughly through all issues/prs soon. Last weeks have been a little crazy with the Laravel 8 release and such.
